# Taller 1

<center>

|     Nombres      |      Apellidos       |     Login     |  Codigo   |
| :--------------: | :------------------: | :-----------: | :-------: |
|     Santiago     | Gonzalez Montealegre | s.gonzalez35  | 202012274 |
| Juliana Carolina |  Cardenas Barragan   | jc.cardenasb1 | 202011683 |

</center>

<br>

---

<br>

---

## El problema de Monty Hall

En un programa concurso Ud puede ganar un carro. Para hacerlo debe seleccionar una de tres puertas.Detras de una de ellas esta el carro y si la selecciona se gana el carro.

Tras su seleccion una de las puertas, el animador del programa abrirá otra de las puertas, detrás de la que hay una cabra. Recuerde que el animador sabe que hay detras de cada puerta.

Tras abrir la otra puerta, el animador le da la posibilidad de cambiar de puerta, ¿Ud se cambia o no de puerta?

---

### 1
Para decidir si debe cambiar o no de puerta Ud ha decidido modelar este problema 
como una Red Bayesiana.

### 2
Defina tres variables aleatorias U, C y A, que representan la puerta seleccionada por Ud, la puerta detras de la cual esta el carro, y la puerta que selecciona el animador, respectivamente. Las tres variables tienen espacio de estados {1, 2, 3}.

### 3

Ud ha definido la siguiente red bayesiana para modelar este problema. En sureporte explique por que decide emplear esta red.

<center>
<img src="https://i.ibb.co/FmyVt7c/Punto1.png" width="400" height="350" />
<img src="https://i.ibb.co/MhpDSGL/Punto2y3.png" width="400" height="350" />
</center>

<br>

**Respuesta**

La puerta que se eligió y la puerta en la que se encuentra el carro son procesos aleatorios entre las tres puertas. Mientras que la puerta que abre el anfitrión esta **condicionada** de la puerta que eligi inicialmente; no puede ser la misma puerta que eligi, y a su vez de la puerta en la que esta ubicada el carro; el anfitrion tiene que abrir una puerta donde detras hay una cabra. Es por esto que se representa la red bayesiana por medio de 2 nodos padres, y un nodo decendiente.

---

### 4
Responda las siguientes preguntas en su reporte:

1. ¿Tiene esta red estructuras en V? Si las tiene, ¿Que implicaciones tienen en la influencia entre las variables de la red?

    **R/**
    
    Esta red **si tiene estructuras en V**. Esta estrucutra esta compuesta por 2 nodos causas; nodos padres, y su efecto comun; nodo descendiente. Aunque los nodos padres en un principio eran marginalmente independientes se vuelven condicionalmente dependientes dada la evidencia de sus hijos. 

1. ¿Que tipo de analisis considera es mas relevante en este caso? ¿Causal, evidencial, intercausal? ¿Por que?

    **R/**
    
    
    En este caso queremos saber si cambio la puerta que seleccione al inicio o no; es decir, queremos saber cual es la puerta que tiene mas probabilidades de que el carro se encuentre detras. Para poder realizar este analisis es necesario contar con un nodo causa y su efecto; la puerta que he seleccionado y la puerta que el anfitrion ha descubierto con la cabra, teniendo estos valores se logra identificar las probabilidades de que el premio se encuentre en cada una de las puertas restantes. Por tal motivo se esta tratando de un **razonamiento intercausal**.

1. ¿Que independencias condicionales captura esta red?

    **R/**
    
    Esta red **no** presenta independecias condicionales, solo independencias marginales.

---

### 5 

El carro ha sido ubicado al azar detras de alguna de las puertas, y Ud selecciona una puerta al azar inicialmente. Modele este comportamiento con las distribuciones de probabilidad P(U) y P(C). 

a. **La puerta que elegi**: 
Las tres puertas tienen la misma probabilidad de ser elegidas

<center>

| U | P(U) | 
| :-: | :-: | 
| 1 | 1/3 |
| 2 | 1/3 |
| 3 | 1/3 |

</center>

<br>

b. **La puerta en la que se encuentra detras el carro**:
El premio puede estar ubicado en cualquiera de las tres puertas

<center>

| C | P(C) | 
| :-: | :-: | 
| 1 | 1/3 |
| 2 | 1/3 |
| 3 | 1/3 |

</center>

---

### 6 

EL comportamiento del Animador (la puerta que decide mostrar) depende tanto de su seleccion como de la puerta donde se encuentra el carro. Modele el comportamiento del Animador con la distribucion condicional de probabilidad $P(A|C,U)$ (note que esta representacion es estocastica por columnas)

<br>

<center>

| P(A &#124; C,U) | C | Puerta 1 | Puerta 1 | Puerta 1 | Puerta 2 | Puerta 2 | Puerta 2 | Puerta 3 | Puerta 3 | Puerta 3 |
| :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: |
| **A** | **U** | **1** | **2** | **3** | **1** | **2** | **3** | **1** | **2** | **3** |
| 1 | | 0 | 0 | 0 | 0 | 0.5 | 1 | 0 | 1 | 0.5 |
| 2 | | 0.5 | 0 | 1 | 0 | 0 | 0 | 1 | 0 | 0.5 |
| 3 | | 0.5 | 1 | 0 | 1 | 0.5 | 0 | 0 | 0 | 0 |

</center>

<br>

---

### 7, 8, 9, 10

In [1]:
# ------------ Warnings of dependencies -----------
import warnings
warnings.filterwarnings("ignore")
# -------------------------------------------------

# 7
#En su editor de python cree un nuevo archivo monty.py e 
#incluya los objetos BayesianNetwork y TabularCPD

from pgmpy.models import BayesianNetwork
from pgmpy.factors.discrete import TabularCPD

# 8
#Defina ahora la estructura de la red incluyendo los arcos y nodos

model = BayesianNetwork([("C","A"),("U","A")])

# 9
#Defina las CPDs de C y U

cpd_c = TabularCPD(variable="C", variable_card=3, values=[[0.33], [0.33], [0.33]])
cpd_u = TabularCPD(variable="U", variable_card=3, values=[[0.33], [0.33], [0.33]])

#10
#Defina la CPD de A

cpd_a = TabularCPD(
    variable="A",
    variable_card=3,
    values=[
        [0, 0, 0, 0, 0.5, 1, 0, 1, 0.5],
        [0.5, 0, 1, 0, 0, 0, 1, 0, 0.5],
        [0.5, 1, 0, 1, 0.5, 0, 0, 0, 0],
    ],
    evidence=["C", "U"],
    evidence_card=[3, 3],
)

---

### 11 

En la documentacion de pgmpy podra encontrar los detalles de la clase TabularCPD, ver https://pgmpy.org/factors/discrete.html. En su reporte justifique el valor de los argumentos usados al crear los tres objetos de la clase TabularCPD.

**Explicación variable cpd_c y cpd_u:**

```py

cpd_c = TabularCPD(variable="C", variable_card=3, values=[[0.33], [0.33], [0.33]])
cpd_u = TabularCPD(variable="U", variable_card=3, values=[[0.33], [0.33], [0.33]])

```


1. El primer parámetro `variable` hace referencia al nombre del objeto cuya distribución de probabilidad condicional vamos a definir; "C" y "U".

2. El segundo parámetro `variable_card` hace referencia al numero de espacio de estados de la variable; en ambos casos se tienen 3 variables de estado: {Puerta 1, Puerta 2, Puerta 3}. 
    
3. El tercer parámetro `values` hace referencia a los valores de la tabla de probabilidad condicional del objeto que vamos a definir; en este caso ambos objetos tienen la misma probabilidad de elegir cualquier puerta $[0.33, 0.33, 0.33]$.

<br>

**Explicación variable cpd_a:**

```py

cpd_a = TabularCPD(
    variable="A", variable_card=3, values=[
        [0, 0, 0, 0, 0.5, 1, 0, 1, 0.5],
        [0.5, 0, 1, 0, 0, 0, 1, 0, 0.5],
        [0.5, 1, 0, 1, 0.5, 0, 0, 0, 0],
    ],
    evidence=["C", "U"], evidence_card=[3, 3],
)

```

1. El primer parámetro `variable` hace referencia al nombre del objeto cuya distribución de probabilidad condicional vamos a definir; "A".

2. El segundo parámetro `variable_card` hace referencia al numero de espacio de estados de la variable; al igual que en los dos casos anteriores se tienen 3 variables de estado: {Puerta 1, Puerta 2, Puerta 3}. 
    
3. El tercer parámetro `values` hace referencia a los valores de la tabla de probabilidad condicional del objeto que vamos a definir; en este caso se hace referencia a una lista.

4. El cuarto parámetro `evidence` hace referencia a las variables CPD que ya han sido definidas y son causales del objeto que vamos a definir.

5. El quinto parámetro `evidence_card` hace referencia al número de espacio de estados que tienen las variables causales del objeto que vamos a definir.

---

### 12, 13, 14, 15

In [2]:
# 12
#Asocie las 3 CPDs a su modelo

model.add_cpds(cpd_c, cpd_u, cpd_a)

#13
#Revise que su modelo este completo

model.check_model()

# 14
#Importe la clase VariableElimination del paquete de inferencia y cree un objeto de
#esta clase

from pgmpy.inference import VariableElimination

infer = VariableElimination(model)

#15
# Suponga que Ud selecciona la puerta 1 y el animador la puerta 3. ¿Cual es la probabilidad 
# de que el carro este detras de cada una de las puertas?

posterior_p = infer.query(variables=["C"], evidence = {"U": 0, "A": 2})
print(posterior_p)

+------+----------+
| C    |   phi(C) |
+======+==========+
| C(0) |   0.3333 |
+------+----------+
| C(1) |   0.6667 |
+------+----------+
| C(2) |   0.0000 |
+------+----------+


---

### 16

Explique el anterior comando a partir de la documentacion del la librerıa https://pgmpy.org/exact_infer/ve.html#pgmpy.inference.ExactInference.VariableElimination.query.

```py

infer.query(variables=["C"], evidence = {"U": 0, "A": 2})

```

1. El primer parámetro `variables` hace referencia a la variable a la cual queremos calcular la probabilidad

2. El segundo parámetro `evidence` hace referencia a un diccionario con pares de valores que contiene las variables y su respectivo espacio de estados que se quiere usar como evidencia para calcular la probabilidad. 

---

### 17 

**Resultados**

<center>

|  **C**   | **phi(c)** |
| :------: | :--------: |
| **C(0)** |    0.33    |
| **C(1)** |    0.66    |
| **C(2)** |     0      |

</center>

Interprete los resultados. ¿Es mejor cambiarse de puerta o no?

<br>

**Respuesta**

Por medio de los resultados podemos asegurar que **es mejor cambiarse de puerta**, ya que de esta manera la probabilidad que tengo de ganar el carro aumenta de 33,33% a 66,67%.

---

### 18 

Modifique la evidencia para considerar otros casos de seleccion de puertas, tanto la que Ud selecciona como la que el Animador abre. ¿Como interpreta estos resultados?

In [3]:
for i in range(0,3):
    for j in range(0,3):
        if i != j:
            prob = infer.query(["C"],evidence={"U": i,"A": j})
            print(f"U = {i} \t A = {j}")
            print(prob)

U = 0 	 A = 1
+------+----------+
| C    |   phi(C) |
+======+==========+
| C(0) |   0.3333 |
+------+----------+
| C(1) |   0.0000 |
+------+----------+
| C(2) |   0.6667 |
+------+----------+
U = 0 	 A = 2
+------+----------+
| C    |   phi(C) |
+======+==========+
| C(0) |   0.3333 |
+------+----------+
| C(1) |   0.6667 |
+------+----------+
| C(2) |   0.0000 |
+------+----------+
U = 1 	 A = 0
+------+----------+
| C    |   phi(C) |
+======+==========+
| C(0) |   0.0000 |
+------+----------+
| C(1) |   0.3333 |
+------+----------+
| C(2) |   0.6667 |
+------+----------+
U = 1 	 A = 2
+------+----------+
| C    |   phi(C) |
+======+==========+
| C(0) |   0.6667 |
+------+----------+
| C(1) |   0.3333 |
+------+----------+
| C(2) |   0.0000 |
+------+----------+
U = 2 	 A = 0
+------+----------+
| C    |   phi(C) |
+======+==========+
| C(0) |   0.0000 |
+------+----------+
| C(1) |   0.6667 |
+------+----------+
| C(2) |   0.3333 |
+------+----------+
U = 2 	 A = 1
+------+--------

**Resultados:**

<center>

|  U = 0   |   A = 1    |   A = 2    |
| :------: | :--------: | :--------: |
|  **C**   | **phi(c)** | **phi(c)** |
| **C(0)** |    0.33    |    0.33    |
| **C(1)** |     0      |    0.66    |
| **C(2)** |    0.66    |     0      |

</center>

<br>

<center>

|  U = 1   |   A = 0    |   A = 2    |
| :------: | :--------: | :--------: |
|  **C**   | **phi(c)** | **phi(c)** |
| **C(0)** |     0      |    0.66    |
| **C(1)** |    0.33    |    0.33    |
| **C(2)** |    0.66    |     0      |

</center>

<br>

<center>

|  U = 2   |   A = 0    |   A = 1    |
| :------: | :--------: | :--------: |
|  **C**   | **phi(c)** | **phi(c)** |
| **C(0)** |     0      |    0.66    |
| **C(1)** |    0.66    |     0      |
| **C(2)** |    0.33    |    0.33    |

</center>

<br>

**Respuesta**

Podemos observar en todos los casos que:

1. **SIEMPRE** va a ser preferible cambiar la puerta que inicialmente escogí. Ya que esta tiene una mayor probabilidad de que atras se encuentre el carro dado que la probabilidad cambia de $33.33 \%$ a $66.67 \%$.

2. La probabilidad de que el premio se encuentre en la puerta que el animador ("A") seleccionó **SIEMPRE** es $0$.

---

### 19

Compare ahora con el caso en que solo tiene como evidencia la puerta que Ud selecciono.

In [4]:
for i in range(0,3):
    prob_2 = infer.query(["C"],evidence={"U": i})
    print("U = "+str(i))
    print(prob_2)

U = 0
+------+----------+
| C    |   phi(C) |
+======+==========+
| C(0) |   0.3333 |
+------+----------+
| C(1) |   0.3333 |
+------+----------+
| C(2) |   0.3333 |
+------+----------+
U = 1
+------+----------+
| C    |   phi(C) |
+======+==========+
| C(0) |   0.3333 |
+------+----------+
| C(1) |   0.3333 |
+------+----------+
| C(2) |   0.3333 |
+------+----------+
U = 2
+------+----------+
| C    |   phi(C) |
+======+==========+
| C(0) |   0.3333 |
+------+----------+
| C(1) |   0.3333 |
+------+----------+
| C(2) |   0.3333 |
+------+----------+


**Resultados:**

<center>

|          |   U = 0    |   U = 1    |   U = 2    |
| :------: | :--------: | :--------: | :--------: |
|  **C**   | **phi(c)** | **phi(c)** | **phi(c)** |
| **C(0)** |    0.33    |    0.33    |    0.33    |
| **C(1)** |    0.33    |    0.33    |    0.33    |
| **C(2)** |    0.33    |    0.33    |    0.33    |

</center>

<br>

**Respuesta**

Tener la información de la puerta que yo escogi no es evidencia suficiente para concluir o determinar si detras de una puerta se encuentra una cabra o un coche, por tal motivo todas las puertas tienen la misma probabilidad de que el premio se encuentre detras de esta.

---

### 20 

Compare ahora con el caso en que solo tiene como evidencia la puerta que abre el Animador

In [5]:
for i in range(0,3):
    prob_3 = infer.query(["C"],evidence={"A": i})
    print("A = "+str(i))
    print(prob_3)

A = 0
+------+----------+
| C    |   phi(C) |
+======+==========+
| C(0) |   0.0000 |
+------+----------+
| C(1) |   0.5000 |
+------+----------+
| C(2) |   0.5000 |
+------+----------+
A = 1
+------+----------+
| C    |   phi(C) |
+======+==========+
| C(0) |   0.5000 |
+------+----------+
| C(1) |   0.0000 |
+------+----------+
| C(2) |   0.5000 |
+------+----------+
A = 2
+------+----------+
| C    |   phi(C) |
+======+==========+
| C(0) |   0.5000 |
+------+----------+
| C(1) |   0.5000 |
+------+----------+
| C(2) |   0.0000 |
+------+----------+


**Resultados:**

<center>

|          |   A = 0    |   A = 1    |   A = 2    |
| :------: | :--------: | :--------: | :--------: |
|  **C**   | **phi(c)** | **phi(c)** | **phi(c)** |
| **C(0)** |     0      |    0.5     |    0.5     |
| **C(1)** |    0.5     |     0      |    0.5     |
| **C(2)** |    0.5     |    0.5     |     0      |

</center>

<br>

**Respuesta**


En el caso en el que solo se conozca la puerta que seleccionó el animador nos quedan 2 posibles opciones: una puerta que detras tienen una cabra o una puerta que detras tiene el carro. Como no se tiene la información de la puerta que he seleccionado, simplemente se puede decir que se reducen las opciones de puertas que puedo escoger a 2 y ambas son equiprobables con respecto a la posibilidad de que al abrirla se encuentre el carro; $50 \%$.

---

### 21

Concluya.

**Respuesta**

Luego de modelar el problema de Monty Hall por medio de una Red Bayesiana se logró concluir que teniendo la información de la puerta que seleccione inicialmente y la puerta que abre el anfitrión, **es preferible cambiar de puerta** ya que la otra puerta tendra el doble de probabilidad de que el carro se encuentre detras.

<br>

---

<br>

---

## El problema de la alarma antirrobo

Ud ha instalado una alarma antirrobo. La alarma es buena detectando intrusos pero tambien se activa con movimientos sısmicos. Ud tiene ademas buena relacion con dos vecinos, Juan y Marıa, quienes lo llaman en caso de que escuchen la alarma. Juan suele llamar al escuchar la alarma, pero tambien se confunde con otros ruidos y llama cuando la alarma no ha sonado. Marıa a veces no escucha la alarma. Su evidencia consiste en si Juan y/o Marıa lo llaman, y basado en esto Ud quiere estimar la probabilidad de que haya habido un robo.

---

### 1 

Para estimar la probabilidad de robo Ud ha decidido modelar este problema como una Red Bayesiana.

### 2 

Defina cinco variables aleatorias R, S, A, J y M que representan el robo, el sismo, la alarma, la llamada de Juan y la de Mar ́ıa, respectivamente. Todas las variables tienen espacio de estados {V, F } (verdadero y falso).

### 3 

Ud ha definido la siguiente red bayesiana para modelar este problema. En su reporte explique por qué decide emplear esta red.

<center>
<img src="https://i.ibb.co/WnZK8Gf/Punto3.png" width="400" height="350" />
<img src="https://i.ibb.co/3sh9HBt/Punto4y5.png" width="400" height="350" />
</center>

<br>

**Respuesta**

La ocurrencia de un intruso entre a la casa o de que se presente un sismo son procesos aleatorios, pero estos son la causa de que se active la alarma. Por otro lado, que llame Juan o Maria se debe a si escuchan que la alarma se activa o no.

---

### 4 

Responda las siguientes preguntas en su reporte:

1. ¿Tiene esta red estructuras en V? Si las tiene, ¿Qué implicaciones tienen en la influencia entre las variables de la red?

    **R/**
    
    Esta red **si tiene estructuras en V**. Esta estrucutra está compuesta por 2 nodos causas; nodos padres, y su efecto comun; nodo descendiente. Robo y Sismo son los nodos explicativos para que la alarma se active o no.

1. ¿Qué tipo de análisis considera es mas relevante en este caso? ¿Causal, evidencial, intercausal? ¿Por qué?

    **R/**

    En este caso queremos saber cual es la probabilidad de que Juan o Maria me llamen dado que se produjo un robo; a partir de los resultados se quiere llegar a las causas. Por tal motivo se esta tratando de un **razonamiento evidencial, explicativo o diagnóstico**.

1. ¿Qué independencias condicionales captura esta red?

    **R/**
    
    Esta red **si** presenta independencias condicionales, se puede observar en el caso de que Juan o Maria llamen dado que sono la alarma. Tener información de que Juan llama no nos da mas información sobre si Maria llama, y viceversa. Este tipo de independencias nos permite factorizar la distribución conjunta.

---

### 5 

La probabilidad de un robo se ha estimado en 0.01, mientras la de un sismo en 0.02. Modele este comportamiento con las distribuciones de probabilidad P(R) y P(S).

a. **Robo**: 

<center>

|  R  | P(R) |
| :-: | :--: |
|  V  | 0.01 |
|  F  | 0.99 |

</center>

b. **Sismo**: 

<center>

|  S  | P(S) |
| :-: | :--: |
|  V  | 0.02 |
|  F  | 0.98 |

</center>

---

### 6

 La alarma responde a un robo o a un sismo de acuerdo con la siguiente CPD:

<center>

| P ( A &#124; R , S) |   R   |   V   |   V   |   F   |   F   |
| :-----------------: | :---: | :---: | :---: | :---: | :---: |
|        **A**        | **S** | **V** | **F** | **V** | **F** |
|          V          |       | 0.95  | 0.94  | 0.29  | 0.001 |
|          F          |       | 0.05  | 0.06  | 0.71  | 0.999 |

</center>

Interprete estas probabilidades en su reporte.

<br>

**Respuesta**


A partir de estas probabilidades se puede concluir que:

1. Es muy probable que la alarma responda ante un robo.
2. Es poco probable que la alarma responda si **no** se presenta un robo y un sismo.
3. Hay mas probabilidades que la alarma **no** responda si solamente se presenta un sismo.

---    

### 7

 Si suena la alarma, Juan llama 9 de cada 10 veces, mientras Marıa llama 7 de cada 10 veces. Si la alarma NO suena, Juan llama 1 de cada 20 veces, mientras María llama 1 de cada 100 veces. Modele este comportamiento con las distribuciones de probabilidad P(J|A) y P(M|A).

<br>

**Respuesta**

a. **Juan**:

<center>

| P( J &#124; A ) |  A  |  A   |
| :-------------: | :-: | :--: |
|      **J**      |  V  |  F   |
|        V        | 0.9 | 0.05 |
|        F        | 0.1 | 0.95 |

</center>

<br>

b. **Maria**:

<center>

| P( M &#124; A ) |  A  |  A   |
| :-------------: | :-: | :--: |
|      **M**      |  V  |  F   |
|        V        | 0.7 | 0.01 |
|        F        | 0.3 | 0.99 |

</center>

---

### 8, 9

In [6]:
# 8

import pgmpy

#En su editor de python cree un nuevo archivo alarma.py e implemente el modelo creado 
# para el problema de la alarma antirrobo. Utilice las clases BayesianNetwork, TabularCPD 
# y VariableElimination de pgmpy.

from pgmpy.models import BayesianNetwork
from pgmpy.factors.discrete import TabularCPD

#Defina ahora la estructura de la red incluyendo los arcos y nodos

model_2 = BayesianNetwork([("R","Al"),("S","Al"),("Al","J"),("Al","M")])

#Defina las CPDs de R y S

cpd_r = TabularCPD(variable="R", variable_card=2, values=[[0.01], [0.99]])
cpd_s = TabularCPD(variable="S", variable_card=2, values=[[0.02], [0.98]])

#Defina la CPD de A

cpd_al = TabularCPD(
    variable="Al",
    variable_card=2,
    values=[
        [0.95, 0.94, 0.29, 0.001],
        [0.05, 0.06, 0.71, 0.999],
    ],
    evidence=["R", "S"],
    evidence_card=[2, 2],
)

#Defina la CPD de J

cpd_j = TabularCPD(
    variable="J",
    variable_card=2,
    values=[
        [0.9, 0.05],
        [0.1, 0.95],
    ],
    evidence=["Al"],
    evidence_card=[2],
)

#Defina la CPD de M

cpd_m = TabularCPD(
    variable="M",
    variable_card=2,
    values=[
        [0.7, 0.01],
        [0.3, 0.99],
    ],
    evidence=["Al"],
    evidence_card=[2],
)

#Asocie las 5 CPDs a su modelo

model_2.add_cpds(cpd_r, cpd_s, cpd_al, cpd_j, cpd_m)

#Revise que su modelo este completo

model_2.check_model()

# 9
#Una vez tenga el modelo construido, obtenga las independencias con el metodo
dependencies = model_2.get_independencies()

# print(len(dependencies))
print(dependencies)


(S ⟂ R)
(S ⟂ M, J | Al)
(S ⟂ J | Al, M)
(S ⟂ M, J | Al, R)
(S ⟂ M | Al, J)
(S ⟂ J | Al, M, R)
(S ⟂ M | Al, R, J)
(J ⟂ M, S, R | Al)
(J ⟂ R, S | Al, M)
(J ⟂ M, R | Al, S)
(J ⟂ M, S | Al, R)
(J ⟂ R | Al, M, S)
(J ⟂ S | Al, M, R)
(J ⟂ M | Al, R, S)
(M ⟂ R, S, J | Al)
(M ⟂ S, J | Al, R)
(M ⟂ R, J | Al, S)
(M ⟂ R, S | Al, J)
(M ⟂ J | Al, R, S)
(M ⟂ S | Al, R, J)
(M ⟂ R | Al, S, J)
(R ⟂ S)
(R ⟂ M, J | Al)
(R ⟂ J | Al, M)
(R ⟂ M, J | Al, S)
(R ⟂ M | Al, J)
(R ⟂ J | Al, M, S)
(R ⟂ M | Al, S, J)


**Respuesta**

Por medio de la formula utilizada se puede comprobar que el modelos presenta independencias marginales y condicionales; 2 y 26 respectivamente. Cada variable cuenta con 7 independencias. Las 2 independencias marginales se presentan en los nodos de robo "R" y sismo "S". 

---

### 10 

Considere 4 casos de evidencia: 

- Recibe llamadas tanto de Juan como de María
- Recibe la llamada de uno solo de ellos
- Ninguno lo llama.

Calcule la probabilidad de que haya ocurrido un robo en cada caso. Incluya sus resultados y análisis en su reporte.


In [7]:

#Importe la clase VariableElimination del paquete de inferencia y cree un objeto de
#esta clase

from pgmpy.inference import VariableElimination

infer_2 = VariableElimination(model_2)

for i in range(0,2):
    for j in range(0,2):
        prob_4 = infer_2.query(["R"],evidence={"J": i, "M": j})
        print("J = "+str(j)+" "+"M = "+str(i))
        print(prob_4)


J = 0 M = 0
+------+----------+
| R    |   phi(R) |
+======+==========+
| R(0) |   0.5565 |
+------+----------+
| R(1) |   0.4435 |
+------+----------+
J = 1 M = 0
+------+----------+
| R    |   phi(R) |
+======+==========+
| R(0) |   0.0484 |
+------+----------+
| R(1) |   0.9516 |
+------+----------+
J = 0 M = 1
+------+----------+
| R    |   phi(R) |
+======+==========+
| R(0) |   0.0634 |
+------+----------+
| R(1) |   0.9366 |
+------+----------+
J = 1 M = 1
+------+----------+
| R    |   phi(R) |
+======+==========+
| R(0) |   0.0009 |
+------+----------+
| R(1) |   0.9991 |
+------+----------+


**Resultados:**

<center>

| phi(R) |  J  |   0    |   0    |   1    |   1    |
| :----: | :-: | :----: | :----: | :----: | :----: |
|   R    |  M  |   0    |   1    |   0    |   1    |
|  R(0)  |     | 0.5565 | 0.0634 | 0.0484 | 0.0009 |
|  R(1)  |     | 0.4435 | 0.9366 | 0.9516 | 0.9991 |

</center>

<br>

**Respuesta**

A partir de estos analisis se puede concluir que:

1. Hay un 56% de probabilidad de que se haya presentado un robo en la vivienda si Juan y Maria me llaman
2. Es muy poco probable que se haya presentado un robo si ninguno de los 2 me llama
3. Si solo me llama uno de los dos, es mas probable que se presente un robo si es Juan el que llama.

---

### 11 

Suponga ahora que, cuando NO se activa la alarma, Juan lo llama con la misma probabilidad de María, es decir, 1 de cada 100 veces. Recalcule sus probabilidades y analice los cambios en su reporte.

In [8]:
#Defina ahora la estructura de la red incluyendo los arcos y nodos

model_3 = BayesianNetwork([("R","Al"),("S","Al"),("Al","Ju"),("Al","M")])

#Defina las CPDs de R y S

cpd_r = TabularCPD(variable="R", variable_card=2, values=[[0.01], [0.99]])
cpd_s = TabularCPD(variable="S", variable_card=2, values=[[0.02], [0.98]])

#Defina la CPD de A

cpd_al = TabularCPD(
    variable="Al",
    variable_card=2,
    values=[
        [0.95, 0.94, 0.29, 0.001],
        [0.05, 0.06, 0.71, 0.999],
    ],
    evidence=["R", "S"],
    evidence_card=[2, 2],
)

#Se redefine la CPD de J

cpd_j2 = TabularCPD(
    variable="Ju",
    variable_card=2,
    values=[
        [0.9, 0.01],
        [0.1, 0.99],
    ],
    evidence=["Al"],
    evidence_card=[2],
)

#Defina la CPD de M

cpd_m = TabularCPD(
    variable="M",
    variable_card=2,
    values=[
        [0.7, 0.01],
        [0.3, 0.99],
    ],
    evidence=["Al"],
    evidence_card=[2],
)

#Asocie las 5 CPDs a su modelo

model_3.add_cpds(cpd_r, cpd_s, cpd_al, cpd_j2, cpd_m)

#Revise que su modelo est ́e completo

model_3.check_model()

#Importe la clase VariableElimination del paquete de inferencia y cree un objeto de
#esta clase

from pgmpy.inference import VariableElimination

infer_3 = VariableElimination(model_3)

for i in range(0,2):
    for j in range(0,2):
        prob_5 = infer_3.query(["R"],evidence={"Ju": i, "M": j})
        print("J = "+str(j)+" "+"M = "+str(i))
        print(prob_5)


J = 0 M = 0
+------+----------+
| R    |   phi(R) |
+======+==========+
| R(0) |   0.5779 |
+------+----------+
| R(1) |   0.4221 |
+------+----------+
J = 1 M = 0
+------+----------+
| R    |   phi(R) |
+======+==========+
| R(0) |   0.1806 |
+------+----------+
| R(1) |   0.8194 |
+------+----------+
J = 0 M = 1
+------+----------+
| R    |   phi(R) |
+======+==========+
| R(0) |   0.0611 |
+------+----------+
| R(1) |   0.9389 |
+------+----------+
J = 1 M = 1
+------+----------+
| R    |   phi(R) |
+======+==========+
| R(0) |   0.0009 |
+------+----------+
| R(1) |   0.9991 |
+------+----------+


**Resultados:**

<center>

| phi(R) |  J  |   0    |   0    |   1    |   1    |
| :----: | :-: | :----: | :----: | :----: | :----: |
|   R    |  M  |   0    |   1    |   0    |   1    |
|  R(0)  |     | 0.5779 | 0.0611 | 0.1806 | 0.0009 |
|  R(1)  |     | 0.4221 | 0.9389 | 0.8194 | 0.9991 |

</center>

<br>

**Respuesta**

Luego de reformular el modelo se puede concluir que:

1. Hay mas probabilidades de que se presente un robo en la vivienda si Juan y Maria me llaman, pasando de un 56% a un 58%.
2. Se tiene la misma probabilidad que la alarma se activara por un robo si ninguno de los 2 me llama.
3. Al contrario que en resultado del numeral anterior, si solo me llama uno de los dos, es mas probable que se presente un robo si es María la que llama.

---